### PR TITLE
Maas 3118 prometheus plugin

### PR DESCRIPTION
**Pull request by tetianakh**
Added a plugin for Prometheus.

The plugin receives a check, retrieves data from Prometheus according to the query provided in the check's source, and returns values extracted from the result.

If request to Prometheus failed, the corresponding graphite metric is incremented. If no data is received (for instance because metric name is misspelled),

### CODE REVIEW

File: alamo_worker/plugins/tests/test_prometheus_plugin.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+
+START = 1507190650
+END = 1507193650
+STEP = 60
+
+
+def get_data():
+    check = {
+        u'environment': u'prod',
```
Comment (deejay1) [inline alamo_worker/plugins/tests/test_prometheus_plugin.py]: why the unicode markers? it's py3 str is by default

File: alamo_worker/plugins/prometheus.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            else:
+                status = RESULT_OK
+                message = None
+            return status, message
+
+    def exception_repr(self, exception: Type[Exception], check_id: int) -> str:
+        if isinstance(exception, PrometheusClientError):
+            return "Prometheus error occured for check {}: '{}'".format(
+                check_id, str(exception))
+        return super(self.__class__, self).exception_repr(exception, check_id)
```
Comment (pnuckowski) [inline alamo_worker/plugins/prometheus.py]: just `super()` without args

File: alamo_worker/plugins/prometheus.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+class PrometheusResult(AbstractResult):
+    __slots__ = ('name', 'values', 'status', 'message')
+
+    def __init__(self, name: str, values: List,
+                 status: int, message: str = None) -> None:
+        self.name = name
+        self.values = Array(values or [])
+        self.status = status
+        self.message = message
```
Comment (pnuckowski) [inline alamo_worker/plugins/prometheus.py]: We also want to have `num_series`

File: alamo_worker/plugins/prometheus.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                message = 'Data not found, check your query'
+                self._monitoring.no_results(data.environment)
+            else:
+                status = RESULT_OK
+                message = None
+            return status, message
+
+    def exception_repr(self, exception: Type[Exception], check_id: int) -> str:
+        if isinstance(exception, PrometheusClientError):
+            return "Prometheus error occured for check {}: '{}'".format(
```
[the remark this answers is not in the document]
Comment (pnuckowski) [reply alamo_worker/plugins/prometheus.py]: it's already logged line: 79 but with exception level. To high ? don't you think?

[the remark this answers is not in the document]
Comment (deejay1) [reply alamo_worker/plugins/prometheus.py]: Nah, exception level is ok as we'd like to have the stacktrace here

File: alamo_worker/plugins/prometheus.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    pass
+
+
+class PrometheusClient(object):
+
+    QUERY_RANGE_ENDPOINT = '/query_range'
+    ERROR_STATUS = 'error'
+
+    def __init__(self):
+        api = settings.PROMETHEUS_URL + settings.PROMETHEUS_API
```
Comment (deejay1) [inline alamo_worker/plugins/prometheus.py]: urljoin here to avoid double / ?

File: alamo_worker/plugins/tests/test_prometheus_plugin.py (hunk 1 above)
Comment (deejay1) [reply alamo_worker/plugins/tests/test_prometheus_plugin.py]: this one is still valid here

File: alamo_worker/plugins/prometheus.py (hunk 5 above)
Comment (tetianakh) [reply alamo_worker/plugins/prometheus.py]: @deejay1 `urljoin` doesn't behave like `+`. For example, `urljoin('http://example.com/a', 'b')` returns `'http://example.com/b'`, and `urljoin('http://example.com/a/b', '/c')` returns `'http://example.com/c'`. I personally find it very confusing and prefer `+`, as the URL and API path are set once in a config file and don't change much (unless we start using a new API version).

**Pull request merged**